doc/adrs/031_separate_metadata_from_representations.adoc

```diff
@@ -0,0 +1,101 @@
+= ADR-031 - Separate metadata representations
+
+== Context
+
+Our representations contain two types of information, some of them are related to the content of the representation and other to its identity.
+When we are manipulating representations, for example to display them in the explorer, we need to manipulate properties related to their identity.
+When we are refreshing a representation, we are mostly updating properties related to its content.
+
+We will split those two set of properties into different objects.
+
+
+== Decision
+
+=== GraphQL API change
+
+The representation metadata will also appear in the GraphQL schema used by Sirius Components in order to be able to request only identity properties for a representation.
+For that, we will introduce an interface named `IRepresentationMetadata` with the following content in the schema:
+
+```
+interface RepresentationMetadata {
+  id: ID!
+  descriptionId: ID!
+  label: String!
+}
+```
+
+With this interface, we wil have various implementations (one for each representation) such as:
+
+```
+type DiagramMetadata implements RepresentationMetadata {
+  id: ID!
+  descriptionId: ID!
+  label: String
+  toolSections: [ToolSection!]!
```

Comment:
Not sure about having the `toolSections` here in the metadata. It does not feel like an intrinsic property of a diagram. Indeed, if/when we add support for layers (in the sense of Sirius Desktop) or similar features, which tools are available at a given time may be dynamic and more part of the "content" of the diagram.

Reply:
Representations will have three source of information:

- Their description with `RepresentationDescription`
- Their metadata with `RepresentationMetadata`
- Their content with `Representation`

Today, we only have `RepresentationDescription` and `Representation` and in order to retrieve the tools of a diagram, you have to load and parse the whole diagram from your database. Computing the tools of the diagram does not involve in any way the content of the diagram today so by moving this field from `Diagram` to `DiagramMetadata` we could stop having to load a diagram three times while opening the EditProjectView:

- One time to find the type of the representation
- One time to load the tools
- One time to load the diagram for the subscription

The first loading will be removed by returning a `DiagramMetadata` from `EditingContext#representation` and the second one would be removed too using the same improvement.

By keeping `toolSections` on `Diagram`, we would need to either return the tool sections with each update of the diagram which would introduce a change in the lifecycle of the tools on the frontend and the backend and it will increase needlessly the bandwidth required to support tools. Since tools are fixed (as of today), they can be computed once for the diagram.

In fact, if we went a step further, the existence of `toolSections` is the only reason motivating the fact that `RepresentationMetadata` is an interface (and thus the lack of kind field) since it needs to be extended in order to add such fields. Since tools are static, we could even move `toolSections` on `DiagramDescription` and have something like this instead on the frontend:

```
getRepresentation {
  viewer {
    project(projectId: "") {
      currentEditingContext {
        representationMetadata(representationId: "") {
          id
          label
          description {
            __typename
            ... on DiagramDescription {
              toolSections {
                tools {
                  id
                }
              }
            }
          }
        }
      }
    }
  }
}
```

This way we could retrieve the metadata of the currently selected representation and, if it's a diagram, its tool sections too. It's a more complex refactoring (not that much more complex) but it may be more elegant. What do you think?

Reply:
:+1: for having `toolSections` on `DiagramDescription`, at least until we have more dynamic tools (which will probably not be before a while).